File: examples/checks/harbor_smr.py
# Independent Python checks for the harbor_smr example.
from .common import run_fragment_checks

CHECKS = [
    ('reserve margin 24 MW exceeds threshold 19 MW', ['PERMIT - North Quay Hydrogen Hub may use https://example.org/insight/harborsmr to run PEM_electrolyzer_train_2 at 16 MW from 2026-06-18T14:00:00Z to 2026-06-18T18:00:00Z.']),
    ('cooling margin 18% exceeds threshold 14%', ['The SMR operator exposes a bounded 18 MW flexible-export insight for day_ahead_balancing, not raw reactor telemetry.']),
    ('no planned outage blocks the balancing window', ['The requested 16 MW electrolysis dispatch fits inside that window, safety margins clear the thresholds, no outage is planned, and the policy permits use only for electrolysis_dispa']),
    ('requested 16 MW fits inside the 18 MW flexible-export insight', ['The approved dispatch is 64 MWh over the four-hour window, scoped to port-hydrogen-hub and PEM_electrolyzer_train_2.']),
    ('serialized insight omits sensitive telemetry terms', ['## Answer']),
    ('aggregate flags confirm raw reactor telemetry stays local', ['## Answer']),
    ('policy permits use for electrolysis dispatch before the insight expires', ['## Answer']),
    ('policy prohibits redistribution for market resale', ['## Answer']),
    ('scope is explicit: device, event, start, and expiry', ['## Answer']),
    ('dispatch plan matches the requested load, power, and insight window', ['## Answer']),
]

def run(ctx):
    return run_fragment_checks(ctx, CHECKS)

File: examples/checks/harbor_smr.go
package main

import (
	"fmt"
	"strings"
)

func checkHarbor(ctx *Context) []Check {
	d := ctx.M()
	agg := asMap(d["Aggregate"])
	insight := asMap(d["Insight"])
	req := asMap(d["Request"])
	dispatch := asMap(d["Dispatch"])
	policy := asMap(d["Policy"])
	th := asMap(d["Thresholds"])
	duration := float64(parseRFC3339ish(str(dispatch["WindowEnd"]))-parseRFC3339ish(str(dispatch["WindowStart"]))) / 3600.0
	energy := num(dispatch["DispatchMW"]) * duration
	serialized := strings.ToLower(str(insight["SerializedLowercase"]))
	sensitive := []string{"coretemperature", "rodposition", "neutronflux", "operatorbadge"}
	omit := true
	for _, t := range sensitive {
		omit = omit && !contains(serialized, t)
	}
	rawLocal := true
	for _, k := range []string{"ContainsCoreTemperature", "ContainsRodPosition", "ContainsNeutronFlux", "ContainsOperatorBadgeIDs"} {
		rawLocal = rawLocal && !boolean(agg[k])
	}
	perm := asMap(policy["Permission"])
	permit := str(d["RequestPurpose"]) == str(req["Purpose"]) && str(req["Purpose"]) == str(perm["Purpose"]) && str(d["RequestAction"]) == str(perm["Action"]) && str(perm["Target"]) == str(insight["ID"]) && str(req["TargetLoad"]) == str(insight["TargetLoad"]) && str(insight["TargetLoad"]) == str(dispatch["ForLoad"]) && num(req["RequestedMW"]) == num(dispatch["DispatchMW"]) && num(req["RequestedMW"]) <= num(insight["ExportMW"]) && num(insight["ExportMW"]) <= num(agg["AvailableFlexibleExportMW"]) && num(agg["ReserveMarginMW"]) >= num(th["MinReserveMarginMW"]) && num(agg["CoolingMarginPct"]) >= num(th["MinCoolingMarginPct"]) && !boolean(agg["PlannedOutage"]) && parseRFC3339ish(str(d["HubAuthAt"])) < parseRFC3339ish(str(insight["ExpiresAt"]))
	pro := asMap(policy["Prohibition"])
	scope := true
	for _, k := range []string{"ScopeDevice", "ScopeEvent", "WindowStart", "ExpiresAt"} {
		scope = scope && str(insight[k]) != ""
	}
	return []Check{{"reserve margin exceeds the configured threshold", num(agg["ReserveMarginMW"]) >= num(th["MinReserveMarginMW"])}, {"cooling margin exceeds the configured threshold", num(agg["CoolingMarginPct"]) >= num(th["MinCoolingMarginPct"])}, {"no planned outage blocks the balancing window", !boolean(agg["PlannedOutage"])}, {"requested dispatch fits inside the flexible-export insight", num(req["RequestedMW"]) <= num(insight["ExportMW"]) && num(insight["ExportMW"]) <= num(agg["AvailableFlexibleExportMW"])}, {"serialized insight omits sensitive reactor telemetry terms", omit}, {"aggregate flags keep raw reactor telemetry local", rawLocal}, {"permission policy authorizes electrolysis dispatch before expiry", permit && contains(ctx.Answer, "PERMIT")}, {"market-resale redistribution is separately prohibited", str(pro["Action"]) == "odrl:distribute" && str(pro["Purpose"]) == "market_resale"}, {"scope is explicit for device, event, start, and expiry", scope}, {"dispatch energy recomputes to 64 MWh over the four-hour window", energy == 64 && contains(ctx.Reason, "64 MWh")}, {"the reported load, power, and window match the request and dispatch", contains(ctx.Answer, str(req["TargetLoad"])) && contains(ctx.Answer, fmt.Sprintf("at %.0f MW", num(req["RequestedMW"]))) && contains(ctx.Answer, str(dispatch["WindowStart"])) && contains(ctx.Answer, str(dispatch["WindowEnd"]))}}
}
